File: Dynamic_Programming/3519. Count Numbers with Non-Decreasing Digits.py
# TC:O(lgn + n * 2 * 2 * 10 * U) U 為前一個數字~10的長度 n = 100 SC:O(n * 2 * 2 * 10 + lgn)
class Solution:
    def countNumbers(self, l: str, r: str, b: int) -> int:
        MOD = 10 ** 9 + 7
        def trans(s: str) -> List[int]:
            x = int(s)
            digits = []
            while x:
                x, r = divmod(x, b)
                digits.append(r)
            digits.reverse()
            return digits

        def check(s:str) -> int:
            m = len(s)
            flag = True
            for i in range(m-1):
                if s[i] > s[i+1]:
                    flag = False
                    break
            return 1 if flag else 0
        l_b = trans(l)
        r_b = trans(r)
        def countNUmber(s:str) -> int:
            n = len(s)
            @cache
            def dfs(i:int, is_num:bool, is_limit:bool, prev:int) -> int:
                if i == n:
                    return 1 if is_num else 0
                ans = 0
                if not is_num:
                    ans = dfs(i + 1, False, False, prev)
                down = prev if is_num else 1# 要比之前的大或者一樣 如果之前都還沒選過 可以從１開始
                up = int(s[i]) if is_limit else b - 1
                for j in range(down, up + 1):
                    ans = (ans + dfs(i + 1, True, is_limit and j == up, j)) % MOD
                return ans
            return dfs(0, False, True, 0) % MOD
        return (countNUmber(r_b) - countNUmber(l_b) + check(l_b) + MOD) % MOD

File: Dynamic_Programming/3519. Count Numbers with Non-Decreasing Digits .cpp
// python 的進制太狗了 故此加上C++版本
class Solution {
    using ll = long long;
    ll MOD = 1e9 + 7;
    string trans(string& s, int b) {
        for (char& c : s) {
            c -= '0';
        }
        string digits;
        while (!s.empty()) {
            string nxt_s; // s / b 的结果
            int rem = 0; // s % b 的结果
            for (char c : s) {
                rem = rem * 10 + c;
                int q = rem / b; // 商
                if (q || !nxt_s.empty()) {
                    nxt_s.push_back(q);
                }
                rem = rem % b;
            }
            digits.push_back(rem);
            s = move(nxt_s);
        }
        ranges::reverse(digits);
        return digits;
    }
    int check(const string &s) {
        int m = s.size();
        for (int i = 0; i < m - 1; i++) {
            if (s[i] > s[i + 1])
                return 0;
        }
        return 1;
    }
public:
    int countNumbers(string l, string r, int b) {
        string l_b = trans(l, b);
        string r_b = trans(r, b);
        auto countNumber = [&](string &s) -> ll {
            int n = s.size();
            vector memo(n, vector<vector<vector<ll>>>(2, vector<vector<ll>>(2, vector<ll>(10, -1))));
            auto dfs = [&](auto &&dfs, int i, bool is_num, bool is_limit, int prev) -> ll {
                if (i == n) return is_num ? 1 : 0;
                ll &res = memo[i][is_num][is_limit][prev];
                if (res != -1) return res;
                res = 0;
                if (!is_num) {
                    res = (res + dfs(dfs, i + 1, 0, 0, prev)) % MOD;
                }
                int up = is_limit ? (s[i]) : (b - 1);
                for (int digit = is_num ? prev : 1; digit <= up; digit++) {
                    res = (res + dfs(dfs, i + 1, 1, is_limit && (digit == up) ? 1 : 0, digit)) % MOD;
                }
                return res;
            };
            return dfs(dfs, 0, false, true, 0) % MOD;
        };
        return (countNumber(r_b) - countNumber(l_b) + check(l_b) + MOD) % MOD;
    }
};
